Oauth2 provider

john-westcott-iv opened the pull request:
Replaces #44
- [x] Update requirements_all.txt
- [x] Add common model changes to prevent time changing from oauth2 system changes
- [x] Add tests
- [x] Fix linting

john-westcott-iv committed:
Add oauth2 provider application

john-westcott-iv committed:
Start unit tests for oauth2 provider

relrod commented on ansible_base/lib/abstract_models/common.py:
```diff
@@ -62,6 +62,11 @@ def save(self, *args, **kwargs):
         update_fields = list(kwargs.get('update_fields', []))
 
+        # If the updating fields lists matches the objects `not_user_modified_fields` then we can just save and return
+        if update_fields == getattr(self, 'not_user_modified_fields', []):
```
:thinking: #308 should make this unnecessary. The field in question is already marked `editable=False`, so that should be enough.

relrod committed:
Regen migrations for common changes

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Tighten up is_external_account() and fix tests

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
fix import alias

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
I regret everything.

Revert "Regen migrations for common changes"

This reverts commit 53b7fb640a84d3c346226d3a5188be48ea467c21.

relrod committed:
Try fixing up migrations while not squashing them

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Start on application tests

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Port a few tests from AWX

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Port another test

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
... And this is why we test :)

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Rework application serializer a bit, client_secret

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Get "related" working for application tokens

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Create an OAuth2Application in demo data

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Add simple tests showing token auth works

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Don't treat Application.client_secret as encrypted

In newer DOT than what AWX uses, Application.client_secret is hashed
automatically with no way to disable that functionality.

There's a PR that allows for disabling that functionality ([0]), but
that hasn't made it into a release.

The DOT hashing is incompatible with our standard encryption - when
DOT gets the value it ends up getting our encrypted string and trying
to act on that. Ideally we'd like to disable their hashing entirely
and use our standard encryption tooling.

AWX avoids this problem by pinning to an older DOT.

For now in DAB we'll just use the upstream hashing, and not treat the
field as an encrypted_fields field to avoid the "double encryption"
issue.

[0]: https://github.com/jazzband/django-oauth-toolkit/pull/1311

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Make is_external_account return the authenticator

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Show the proper authenticator type in error

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Welp, that took a while.

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Finish test_token tests

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Tidy up another test file

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Update fixture tuple application fixtures return

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Track oauth models in activity stream and sanitize

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Provide view-level hook for extra related fields

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Start on /users/PK/<oauth stuff>/ endpoints

Particularly, start on /users/PK/personal_tokens/

No tests, yet.

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Some coverage for the PAT mixin hack

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Nix updated/created from token model

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Nix DOT updated/created fields from some models

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Use the API for the token fixture

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Clean up migrations

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Start a doc for differences from AWX

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Make summary fields useful for Application

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Get /users/N/authorized_tokens/ working

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Just hardcode the user model basename in the mixin

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
We can't check user actions when we add more

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Finish porting AWX tests over

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Use CommonModel fields for created and modified

We have to use the right order of inheritance for this to work, and
not null out fields that we actually want :)

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod commented on ansible_base/oauth2_provider/models/access_token.py:
```diff
@@ -0,0 +1,93 @@
+            if external_account:
+                raise oauth2.AccessDeniedError(
+                    _('OAuth2 Tokens cannot be created by users associated with an external authentication provider (%(authenticator)s)')
+                    % {'authenticator': external_account.type}
```
Type might not be unique, so use the authenticator's name instead

john-westcott-iv commented on ansible_base/oauth2_provider/models/application.py:
```diff
@@ -0,0 +1,88 @@
+    activitystream = AuditableModel
+
+
+DATA_URI_RE = re.compile(r'.*')  # FIXME
```
Can we tighten this up a bit? It looks like this is how it is in AWX so I'm fine with merging this as is for now.

relrod replied:
https://github.com/ansible/django-ansible-base/issues/372

john-westcott-iv commented on ansible_base/lib/serializers/common.py:
```diff
@@ -47,10 +47,16 @@ def is_list_view(self) -> bool:
     def _get_related(self, obj) -> dict[str, str]:
         if obj is None:
             return {}
+        related_fields = {}
+        view = self.context.get('view')
+        if view is not None and hasattr(view, 'extra_related_fields'):
+            related_fields.update(view.extra_related_fields(obj))
```
Did we have another pattern for extra_related_views somewhere?
If so are we going to consolidate on this solution?

relrod replied:
We have several, but they are at different points in the pipeline. We can extend from the model level, and we can override `_get_related()` in the serializer, but there was nothing at the view/viewset level, so this adds that.

john-westcott-iv commented on ansible_base/lib/utils/views/ansible_base.py:
```diff
@@ -51,3 +51,13 @@ def finalize_response(self, request, response, *args, **kwargs):
 
         return response
+
+    def extra_related_fields(self, obj):
+        """
+        A hook for adding extra related fields to serializers which
+        make use of this view/viewset.
+
+        This is particularly useful for mixins which want to extend a viewset
+        with additional actions and provide those actions as related fields.
+        """
+        return {}
```
Should we document this somewhere? Maybe in https://github.com/ansible/django-ansible-base/blob/devel/docs/lib/views.md?

relrod replied:
Filed https://github.com/ansible/django-ansible-base/issues/371

john-westcott-iv commented on ansible_base/oauth2_provider/models/access_token.py:
```diff
@@ -0,0 +1,93 @@
+        ordering = ('id',)
+        swappable = "OAUTH2_PROVIDER_ACCESS_TOKEN_MODEL"
+
+    user = models.ForeignKey(
+        settings.AUTH_USER_MODEL,
+        on_delete=models.CASCADE,
+        blank=True,
+        null=True,
+        related_name="access_tokens",
+        help_text=_('The user representing the token owner'),
+    )
```
General comment here, should we try and grab the fields we are just extending from the parent class and then add our stuff in (like related_name/help_text). If its too much of a pain we can go with this.

relrod replied:
_Can_ we do that? I know django does a _lot_ of metaclass hackery with model classes, and what you end up getting as a "model" is vastly different than what it's actually defined as. We'd end up grabbing the fields after all that metaclass logic has run, so I don't know if that works or not. :thinking:

john-westcott-iv replied:
Thats a good point, maybe we don't want to do that. We can explore this later as I am sure we do similar things in other places.

john-westcott-iv commented on ansible_base/oauth2_provider/models/access_token.py:
```diff
@@ -0,0 +1,93 @@
+        if valid:
+            self.last_used = now()
+
+            def _update_last_used():
+                if OAuth2AccessToken.objects.filter(pk=self.pk).exists():
+                    self.save(update_fields=['last_used'])
+
+            connection.on_commit(_update_last_used)
```
In looking at this now I am wondering why AWX did it this way? This seems like it would only save the last used of a token if something made a commit to the DB and maybe not a read? Either way, this is non-blocking for this PR.

relrod replied:
This is in `is_valid()` which is what DOT uses to check if the token is expired or not, so this seems like a reasonable place for this (though it could probably stand to be commented what's going on).

Basically DOT calls this method to see if it's still a valid token, then we say "if it is valid -- on the next commit bump the `last_used` because it was accessed".

john-westcott-iv commented on requirements/requirements_oauth2_provider.in:
```diff
@@ -0,0 +1 @@
+django-oauth-toolkit>=1.7.1 # This is pinned this way so that DAB is compatible with AWX
```
We should be able to just remove this pin now right?

relrod replied:
Yep!

john-westcott-iv commented on pyproject.toml:
```diff
@@ -72,7 +74,7 @@ force-exclude = '''
 profile = "black"
 line_length = 160
-extend_skip = [ "ansible_base/authentication/migrations", "ansible_base/activitystream/migrations", "test_app/migrations" ]
+extend_skip = [ "ansible_base/authentication/migrations", "ansible_base/activitystream/migrations", "test_app/migrations", "ansible_base/oauth2_provider/migrations" ]
```
Is it possible to just `ansible_base/**/migrations` this?
Non blocking.

john-westcott-iv commented on docs/apps/oauth2_provider.md:
```diff
@@ -0,0 +1,6 @@
+# Differences from AWX
+
+* Because of how DAB's router works, we don't allow for POSTing to (for example)
+  `/applications/PK/tokens/` to create a token which belongs to an application.
+  The workaround is to just use /tokens/ and in the body specify
+  `{"application": PK}`.
```
If we can pull in any AWX docs here that would be 👍

relrod replied:
https://github.com/ansible/django-ansible-base/issues/373

john-westcott-iv commented on ansible_base/oauth2_provider/models/application.py:
```diff
@@ -0,0 +1,88 @@
+        ("authorization-code", _("Authorization code")),
+        ("password", _("Resource owner password-based")),
+    )
+
```
Do we want to do `prevent_search` on the an applications client_id (from the Abstract model)?
```
    * :attr:`client_id` The client identifier issued to the client during the
                        registration process as described in :rfc:`2.2`
```

relrod replied:
The `client_id` isn't secret, right? Just the `client_secret`?

john-westcott-iv replied:
Yea, thats true, we usually leave the id as not encrypted.

john-westcott-iv commented on ansible_base/oauth2_provider/models/id_token.py:
```diff
@@ -0,0 +1,19 @@
+    activitystream = AuditableModel
+
+
+class OAuth2IDToken(CommonModel, oauth2_models.AbstractIDToken, activitystream):
```
Do we need/want prevent search on the `jti` field from AbstractIDToken?
```
    * :attr:`jti` ID token JWT Token ID, to identify an individual token
```

This would also apply to the other tokens as well if we are going to do this.

relrod replied:
I looked into this when I added `prevent_search()` on the access token and I _think_ the JTI isn't secret? It's just a UUID that identifies a token, but not a token itself. It would be nice to have someone more familiar clarify that though.

john-westcott-iv replied:
From a quick scan of the awx codebase it looks like they don't do anything with the jti field so we are fine to leave this as is.

john-westcott-iv commented on ansible_base/oauth2_provider/models/refresh_token.py:
```diff
@@ -0,0 +1,23 @@
+from ansible_base.lib.abstract_models.common import CommonModel
+from ansible_base.lib.utils.models import prevent_search
+
+activitystream = object
```
Random realization, I'm surprised this isn't Object instead of object.

relrod replied:
Well, Python built-in types are lowercase (except exceptions :sweat_smile:), and `object` is just another built-in type.

john-westcott-iv commented on ansible_base/oauth2_provider/models/refresh_token.py:
```diff
@@ -0,0 +1,23 @@
+
+class OAuth2RefreshToken(CommonModel, oauth2_models.AbstractRefreshToken, activitystream):
+    class Meta(oauth2_models.AbstractRefreshToken.Meta):
+        verbose_name = _('access token')
```
`refresh token`?

john-westcott-iv commented on ansible_base/oauth2_provider/serializers/application.py:
```diff
@@ -0,0 +1,92 @@
+            elif request.method == 'POST':
+                if self.oauth2_client_secret is None:
+                    # This should be an impossible case, but...
+                    return ENCRYPTED_STRING
```
Is this right? Would we want to return `ENCRYPTED_STRING` if we didn't have a secret? Seems like it could mask an issue.

relrod replied:
I removed this case, so now it'll just show the `null` if it's `None`.

john-westcott-iv commented on ansible_base/oauth2_provider/serializers/application.py:
```diff
@@ -0,0 +1,92 @@
+        return ret
+
+    def _summary_field_tokens(self, obj):
+        token_list = [{'id': x.pk, 'token': ENCRYPTED_STRING, 'scope': x.scope} for x in obj.access_tokens.all()[:10]]
```
Pulling in just the first 10 tokens feels weird. But I guess its what AWX does.

relrod replied:
Yeah that felt weird to me too. My guess is that it was for performance reasons, but it feels like in practice someone wouldn't have hundreds/thousands of keys anyway. But I kept it because it's what AWX did.

john-westcott-iv commented on ansible_base/oauth2_provider/serializers/token.py:
```diff
@@ -0,0 +1,101 @@
+from ansible_base.lib.utils.settings import get_setting
+from ansible_base.oauth2_provider.models import OAuth2AccessToken, OAuth2RefreshToken
+
+logger = logging.getLogger("ansible_base.serializers.oauth2_provider")
```
`token` instead of `oauth2_provider`.

john-westcott-iv commented on ansible_base/oauth2_provider/serializers/token.py:
```diff
@@ -0,0 +1,101 @@
+class BaseOAuth2TokenSerializer(CommonModelSerializer):
+    refresh_token = SerializerMethodField()
+    token = SerializerMethodField()
+    ALLOWED_SCOPES = ['read', 'write']
```
Should we pull this out of the scope field from the model so that if we ever changed that this would auto-magically update?

john-westcott-iv commented on ansible_base/oauth2_provider/serializers/token.py:
```diff
@@ -0,0 +1,101 @@
+
+class OAuth2TokenSerializer(BaseOAuth2TokenSerializer):
+    def create(self, validated_data):
+        current_user = self.context['request'].user
```
This should be fine but should we use `get_current_user` from `django.crum` here?

john-westcott-iv commented on ansible_base/oauth2_provider/views/token.py:
```diff
@@ -0,0 +1,57 @@
+
+
+class TokenView(oauth_views.TokenView, AnsibleBaseDjangoAppApiView):
+    # There is a big flow of logic that happens around this behind the scenes.
+    #
+    # oauth2_provider.views.TokenView inherits from oauth2_provider.views.mixins.OAuthLibMixin
+    # That's where this method comes from originally.
+    # Then *that* method ends up calling oauth2_provider.oauth2_backends.OAuthLibCore.create_token_response
+    # Then *that* method ends up (ultimately) calling oauthlib.oauth2.rfc6749....
```
This applies to the create_token_response method or the class? Can we specify whatever it is instead of the term `this`?

relrod committed:
Nix dependency LB

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Use wildcard

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
s/access/refresh/

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Fix logger path

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
DRY `ALLOWED_SCOPES` in token serializer

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Use crum's get_current_user()

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Be specific because people aren't mind-readers

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Use authenticator name instead of its type

Signed-off-by: Rick Elrod <rick@elrod.me>

relrod committed:
Just show the None if it's None

Signed-off-by: Rick Elrod <rick@elrod.me>
